democluster: cockroach demo panics whenever the demo cluster fails to start

**Describe the problem**

`./cockroach demo` panics whenever the demo cluster fails to start

```
$ cockroach git:(master) ./cockroach demo
*
* ERROR: listen tcp 127.0.0.1:26257: bind: address already in use
*
*
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.sigpanic
*   | 	/Users/jaylim/.go/1.17.5/src/runtime/signal_unix.go:735
*   | github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).AnnotateCtx
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1223
*   | github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).Start.func8.1
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:419
*   | github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:107
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:563
*   | github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).Close
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:806
*   | github.com/cockroachdb/cockroach/pkg/cli.runDemo
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:246
*   | github.com/cockroachdb/cockroach/pkg/cli.init.5.func1
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:57
*   | github.com/cockroachdb/cockroach/pkg/cli/clierrorplus.MaybeDecorateError.func1
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/clierrorplus/decorate_error.go:67
*   | github.com/spf13/cobra.(*Command).execute
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
*   | github.com/spf13/cobra.(*Command).ExecuteC
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:974
*   | github.com/spf13/cobra.(*Command).Execute
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:902
*   | github.com/cockroachdb/cockroach/pkg/cli.Run
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:295
*   | github.com/cockroachdb/cockroach/pkg/cli.doMain
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:137
*   | github.com/cockroachdb/cockroach/pkg/cli.Main
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:64
*   | main.main
*   | 	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach-short/main.go:26
*   | runtime.main
*   | 	/Users/jaylim/.go/1.17.5/src/runtime/proc.go:255
*   | runtime.goexit
*   | 	/Users/jaylim/.go/1.17.5/src/runtime/asm_amd64.s:1581
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
*
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7046c7e]

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.RecoverAndReportPanic({0x9b33fb8, 0xc0001a4048}, 0xc000d79b00)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:124 +0x78
panic({0x809e2c0, 0xc8bf450})
	/Users/jaylim/.go/1.17.5/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).AnnotateCtx(0x0, {0x9b33fb8, 0xc0001a4048})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1223 +0x1e
github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).Start.func8.1()
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:419 +0x48
github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0xc000d79b00)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:107 +0x1a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc000d79b00, {0x9b33fb8, 0xc0001a4048})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:563 +0x262
github.com/cockroachdb/cockroach/pkg/cli/democluster.(*transientCluster).Close(0xc0001842d0, {0x9b33fb8, 0xc0001a4048})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/democluster/demo_cluster.go:806 +0x34
github.com/cockroachdb/cockroach/pkg/cli.runDemo(0xc911900, {0x0, 0x0})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:246 +0xec6
github.com/cockroachdb/cockroach/pkg/cli.init.5.func1(0xc0017c3ae0, {0x40b5364, 0x0, 0x7})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:57 +0x1d
github.com/cockroachdb/cockroach/pkg/cli/clierrorplus.MaybeDecorateError.func1(0xc911900, {0xc9ac320, 0x0, 0x0})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/clierrorplus/decorate_error.go:67 +0x43
github.com/spf13/cobra.(*Command).execute(0xc911900, {0xc9ac320, 0x0, 0x0})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc909e80)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:902
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:295
github.com/cockroachdb/cockroach/pkg/cli.doMain(0xc911900, {0xc00081c3ea, 0x4})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:137 +0x1e9
github.com/cockroachdb/cockroach/pkg/cli.Main()
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:64 +0x136
main.main()
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/cmd/cockroach-short/main.go:26 +0x17
```

**To Reproduce**

`./cockroach start-single-node` in one terminal, and `./cockroach demo` in another.

**Expected behavior**

When the demo cluster fails to start, it should exit cleanly.

**Additional data**

I believe this is a regression from https://github.com/cockroachdb/cockroach/pull/72924.

`ts` here will never be nil: https://github.com/cockroachdb/cockroach/blob/4ea464d8376402e55bf85d4e69efd4abba33acdd/pkg/cli/democluster/demo_cluster.go#L418-L420

If the test cluster fails to start up, we still return something; it's just that the SQLServer field is nil: https://github.com/cockroachdb/cockroach/blob/4ea464d8376402e55bf85d4e69efd4abba33acdd/pkg/server/testserver.go#L728

**Environment:**

This is cockroachdb on the `master` branch, running on Mac:

```
$  cockroach git:(master) ./cockroach version
Build Tag:        v22.1.0-alpha.1-914-g4ea464d837
Build Time:
Distribution:     CCL
Platform:         darwin amd64 (x86_64-apple-darwin20.6.0)
Go Version:       go1.17.5
C Compiler:       Apple LLVM 13.0.0 (clang-1300.0.29.30)
Build Commit ID:  4ea464d8376402e55bf85d4e69efd4abba33acdd
Build Type:       development
```

cc: @andreimatei

--- a/pkg/cli/democluster/demo_cluster.go
+++ b/pkg/cli/democluster/demo_cluster.go
@@ -392,14 +392,14 @@ func (c *transientCluster) Start(
 			for i := 0; i < c.demoCtx.NumNodes; i++ {
 				latencyMap := c.servers[i].Cfg.TestingKnobs.Server.(*server.TestingKnobs).ContextTestingKnobs.ArtificialLatencyMap
 				c.infoLog(ctx, "starting tenant node %d", i)
-				stopper := stop.NewStopper()
+				tenantStopper := stop.NewStopper()
 				ts, err := c.servers[i].StartTenant(ctx, base.TestTenantArgs{
 					// We set the tenant ID to i+2, since tenant 0 is not a tenant, and
 					// tenant 1 is the system tenant. We also subtract 2 for the "starting"
 					// SQL/HTTP ports so the first tenant ends up with the desired default
 					// ports.
 					TenantID:         roachpb.MakeTenantID(uint64(i + 2)),
-					Stopper:          stopper,
+					Stopper:          tenantStopper,
 					ForceInsecure:    c.demoCtx.Insecure,
 					SSLCertsDir:      c.demoDir,
 					StartingSQLPort:  c.demoCtx.SQLPort - 2,
@@ -418,7 +418,7 @@ func (c *transientCluster) Start(
 					if ts != nil {
 						stopCtx = ts.AnnotateCtx(stopCtx)
 					}
-					stopper.Stop(stopCtx)
+					tenantStopper.Stop(stopCtx)
 				}))
 				if err != nil {
 					return err

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -725,7 +725,10 @@ func (ts *TestServer) StartTenant(
 		baseCfg,
 		sqlCfg,
 	)
-	return &TestTenant{SQLServer: sqlServer, Cfg: &baseCfg, sqlAddr: addr, httpAddr: httpAddr}, err
+	if err != nil {
+		return nil, err
+	}
+	return &TestTenant{SQLServer: sqlServer, Cfg: &baseCfg, sqlAddr: addr, httpAddr: httpAddr}, nil
 }
 
 // ExpectedInitialRangeCount returns the expected number of ranges that should

